src/pulsar_socket_writer.erl

```diff
@@ -0,0 +1,147 @@
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+
+-module(pulsar_socket_writer).
+
+-behaviour(gen_server).
+
+%% API
+-export([
+    start_link/3,
+    stop/1,
+
+    send_batch_async/7,
+    ping_async/1,
+    pong_async/1
+]).
+
+%% `gen_server' API
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+%%------------------------------------------------------------------------------
+%% Type declarations
+%%------------------------------------------------------------------------------
+
+%% Calls/Casts/Infos
+-record(send_batch, {topic, encoded_packet, num_messages}).
+-record(ping, {}).
+-record(pong, {}).
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+start_link(Host, Port, Opts) ->
+    OwnerPid = self(),
+    case pulsar_socket:connect(Host, Port, Opts) of
+        {ok, Sock} ->
+            Params = #{owner => OwnerPid, sock => Sock, opts => Opts},
+            case gen_server:start_link(?MODULE, Params, []) of
+                {ok, SockPid} ->
+                    {ok, {SockPid, Sock}};
+                Error ->
+                    pulsar_socket:close(Sock, Opts),
+                    Error
+            end;
+        Error ->
+            Error
+    end.
+
+stop(SockPid) ->
+    try
+        gen_server:stop(SockPid)
+    catch
+        exit:noproc ->
+            ok
+    end.
+
+send_batch_async(SockPid, Topic, Messages, SequenceId, ProducerId, ProducerName, Opts) ->
+    {NumMessages, EncodedPacket0} =
+        pulsar_socket:encode_send_batch_message_packet(Messages, SequenceId,
+                                                       ProducerId, ProducerName, Opts),
+    EncodedPacket = iolist_to_binary(EncodedPacket0),
+    Req = #send_batch{topic = Topic, encoded_packet = EncodedPacket, num_messages = NumMessages},
+    safe_cast(SockPid, Req).
+
+ping_async(SockPid) ->
+    safe_cast(SockPid, #ping{}).
+
+pong_async(SockPid) ->
+    safe_cast(SockPid, #pong{}).
+
+%%------------------------------------------------------------------------------
+%% `gen_server' API
+%%------------------------------------------------------------------------------
+
+init(#{sock := Sock, owner := OwnerPid, opts := Opts}) ->
+    logger:set_process_metadata(#{domain => [pulsar, socket]}),
+    set_label({pulsar_socket, OwnerPid}),
```

Comment:
owner pid is only used as process label ?
maybe pass down owner's label

after all owner's pid can be found in `process_info(Pid, links)`

Reply:
:checkered_flag: